pkg/server/io/helpers_windows.go

```diff
@@ -25,52 +25,111 @@ import (
 
 	winio "github.com/Microsoft/go-winio"
 	"github.com/containerd/containerd/cio"
-	"github.com/containerd/containerd/log"
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 )
 
+// delayedConnection is a io.ReadWriteCloser that takes a net.Listener, calls Accept on it, and
+// then blocks any Read/Write/Close operations until a connection has been established. It is
+// created by calling newDelayedConnection, and the context.Context passed to that function can
+// be canceled to abort the Accept call.
 type delayedConnection struct {
-	l    net.Listener
-	con  net.Conn
-	wg   sync.WaitGroup
-	once sync.Once
+	con net.Conn
+	// listenDoneCh is closed when listening is complete. This means the delayedConnection
+	// will either be in a connected state, or an error state.
+	listenDoneCh chan struct{}
+	listenErr    error
+	// closeCh is closed to indicate to waitListener that listening should be aborted.
+	closeCh chan struct{}
 }
 
 func (dc *delayedConnection) Write(p []byte) (int, error) {
-	dc.wg.Wait()
-	if dc.con != nil {
-		return dc.con.Write(p)
+	<-dc.listenDoneCh
+	if dc.listenErr != nil {
+		return 0, errors.Wrap(dc.listenErr, "connection failed")
 	}
-	return 0, errors.New("use of closed network connection")
+	return dc.con.Write(p)
 }
 
 func (dc *delayedConnection) Read(p []byte) (int, error) {
-	dc.wg.Wait()
-	if dc.con != nil {
-		return dc.con.Read(p)
+	<-dc.listenDoneCh
+	if dc.listenErr != nil {
+		return 0, errors.Wrap(dc.listenErr, "connection failed")
 	}
-	return 0, errors.New("use of closed network connection")
+	return dc.con.Read(p)
 }
 
-func (dc *delayedConnection) unblockConnectionWaiters() {
-	defer dc.once.Do(func() {
-		dc.wg.Done()
-	})
+// waitListener runs as a separate goroutine and manages the listener
+// for the delayedConnection.
+func (dc *delayedConnection) waitListener(ctx context.Context, l net.Listener) {
+	// We block this entire function on Accept below, and only unblock when either the
+	// listener successfully connects, or if an error is encountered.
+	// We can force unblock the Accept by closing the listener, which is done by the
+	// goroutine below in several cases.
+	ch := make(chan struct{})
+	go func() {
+		select {
+		case <-ctx.Done(): // Context hit deadline or canceled.
+		case <-dc.closeCh: // Close was called.
+		case <-ch: // Accept finished properly.
+		}
+		l.Close()
+	}()
+	con, acceptErr := l.Accept()
+	// Unblock the goroutine above. It's okay for it to close the listener now, we don't need it
+	// anymore.
+	close(ch)
+	if err := ctx.Err(); err != nil {
+		// If the context was canceled, use its error instead of the one from Accept. This ensures
+		// Read/Write will return an error like "connection failed: context deadline exceeded".
+		dc.listenErr = err
+		if con != nil {
+			dc.con.Close()
```

Comment:
Wouldn't `dc.con` not be assigned to anything in this case? I only see it being assigned in the else case below. Is this supposed to be `con.Close()`?

Reply:
You're right, this is a typo